src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -112,22 +112,46 @@ uct_ib_mlx5_devx_reg_ksm(uct_ib_mlx5_md_t *md, int atomic, uint64_t address,
     return UCS_OK;
 }
 
+static void uct_ib_mlx5_devx_ksm_log(uct_ib_mlx5_md_t *md, int atomic,
+                                     void *address, size_t length,
+                                     uint64_t iova, ucs_status_t status, int mt)
+{
+    ucs_debug("KSM %s-thread memory registration status \"%s\" "
+              "range %p..%p iova %p%s on %s",
+              mt ? "multi" : "single", ucs_status_string(status), address,
+              UCS_PTR_BYTE_OFFSET(address, length), (void*)iova,
+              atomic ? " atomic" : "", uct_ib_device_name(&md->super.dev));
+}
+
+/*
+ * Register KSM data for given memory handle. Can work only with MRs
+ * that were registered in multi-threaded mode.
+ */
 static ucs_status_t
-uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md, int atomic, void *address,
-                              uct_ib_mlx5_devx_ksm_data_t *ksm_data,
-                              size_t length, uint64_t iova, uint32_t mkey_index,
-                              const char *reason, struct mlx5dv_devx_obj **mr_p,
-                              uint32_t *mkey)
+uct_ib_mlx5_devx_reg_ksm_data_mt(uct_ib_mlx5_md_t *md, int atomic,
+                                 void *address,
+                                 uct_ib_mlx5_devx_ksm_data_t *ksm_data,
+                                 uint32_t iova_offset, uint32_t mkey_index,
+                                 const char *reason,
+                                 struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
 {
-    void *mr_address = address;
+    struct ibv_mr *last_mr = ksm_data->mrs[ksm_data->mr_num - 1];
+    uint64_t iova          = (uint64_t)address + iova_offset;
+    void *mr_address       = address;
+    size_t list_size       = ksm_data->mr_num;
     ucs_status_t status;
     char *in;
     void *klm;
     int i;
 
-    status = uct_ib_mlx5_alloc_mkey_inbox(ksm_data->mr_num, &in);
+    /* Add offset to workaround CREATE_MKEY range check issue */
+    if (iova_offset > 0) {
+        ++list_size;
+    }
```

Comment:
can we share the same code to workaround single and multi threaded cases?

Reply:
As I understand, this `ucs_div_round_up` covers also another workaround (for alignment) for single threaded registration case. MT registration still doesn't work in that case. 

So it may be complicated to unify them in this patch. Maybe later, when MT registration for unaligned buffers would work, it can be possible.